jobclient/python/jobs.py

```diff
@@ -0,0 +1,381 @@
+# Copyright (c) Two Sigma Open Source, LLC
+#
+# Licensed under the Apache license, Version 2.0 (the "License");
+# you may not use this file ecept in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+
+import constraints
+import util
+
+from copy import deepcopy
+from enum import Enum
+from uuid import UUID
+from typing import Dict, List, Optional, Set
+
+from constraints import Constraint
+from instance import Executor, Instance
+from util import FetchableUri
+
+
+class Status(Enum):
+    """
+    The curent status of a job.
+
+    A job in Cook has only three statuses: WAITING, RUNNING, COMPLETED. The
+    INITIALIZED status is for a `JobListener` to receive status updates.
+    """
+    INITIALIZED = 'INITIALIZED'
+    WAITING = 'WAITING'
+    RUNNING = 'RUNNING'
+    COMPLETED = 'COMPLETED'
+
+    def __str__(self):
+        return self.value
+
+    def __repr__(self):
+        return f'Status.{self.value}'
+
+    @staticmethod
+    def from_string(name: str) -> 'Status':
+        """Parse a `Status` from a case-invariant string representation."""
+        return _JOB_STATUS_LOOKUP[name.lower()]
+
+
+_JOB_STATUS_LOOKUP = {
+    'initialized': Status.INITIALIZED,
+    'waiting': Status.WAITING,
+    'running': Status.RUNNING,
+    'completed': Status.COMPLETED
+}
+
+
+class State(Enum):
+    """The current state of a job.
+
+    Indicates whether a job is currently running, has finished and succeeded,
+    or has finished and failed.
+    """
+    WAITING = 'WAITING'
+    PASSED = 'PASSED'
+    FAILED = 'FAILED'
+
+    def __str__(self):
+        return self.value
+
+    def __repr__(self):
+        return f'State.{self.value}'
+
+    @staticmethod
+    def from_string(name: str) -> 'State':
+        """Parse a `State` from a case-invariant string representation."""
+        return _JOB_STATE_LOOKUP[name.lower()]
+
+
+_JOB_STATE_LOOKUP = {
+    'waiting': State.WAITING,
+    'passed': State.PASSED,
+    'failed': State.FAILED
+}
+
+
+class Application:
+    name: str
+    version: str
+
+    def __init__(self, name: str, version: str):
+        self.name = name
+        self.version = version
+
+    def to_dict(self) -> dict:
+        return {
+            'name': self.name,
+            'version': self.version
+        }
+
+    @classmethod
+    def from_dict(cls, d: dict) -> 'Application':
+        return cls(**d)
+
+
+class Job:
+    """A job object returned from Cook."""
+    command: str
+    mem: float
+    cpus: float
+
+    uuid: UUID
+    name: str
+    max_retries: int
+    max_runtime: int
```

Comment:
Can we convert this to a Python timedelta?

Reply:
(Same for the other time-related fields below)

Reply:
:+1: